### PR TITLE
Deinline printing functions

### PR DESCRIPTION
**Pull request by albinahlback**
@fredrik-johansson can you tell me what you think of such a layout? Only a draft än så länge.

Most notably, no printer function should be inlined. Printer functions is not functions where one usually requires performance, but rather requires a smaller size.

Moreover, I think it is nice to have one file for all I/O, including simple setters and getters for strings as they are very connected to the printer functions.

Later on, I would like to switch from `flint_printf` to `printf` internally as `flint_printf` obviously is slower. Ideally, every printer function would only print to a file stream once, as I/O calls are quite expensive and multithreading can mess up the order if printers are called in parallel; better to build up a string and then call `fprintf` once. In this case it is nice if string functions and I/O are bundled together. And it also reduces the number of files.

### CODE REVIEW

**Comment by fredrik-johansson** [comment]
> Ideally, every printer function would only print to a file stream once, as I/O calls are quite expensive and multithreading can mess up the order if printers are called in parallel

This is not a good idea. When you have computed a huge object that pushes the limits of your RAM, you want to be able to write it to a file without building a string (which may be even larger than the original object) in memory.

The stream interface in Calcium / GR is designed to simplify IO by allowing the same code for get_str and (f)print methods, and I had in mind using it more widely in FLINT. Of course, you could do some buffering inside such a stream if it improves file IO. But I think typical fprintf implementations are buffered these days, so are you sure doing it yourself would be significantly faster?

File: src/flint.h [hunk 1]
```diff
@@ -12,14 +12,21 @@
...
+
 #if !defined(_MSC_VER)
 #include <sys/param.h> /* for BSD define */
 #endif
-#include <stdio.h>
-#include <stdarg.h>
 #include <gmp.h>
 #include "limits.h"
 #include "longlong.h"
+#include "flint_types.h"
```
Comment (fredrik-johansson) [inline src/flint.h:29]: Isn't the inclusion of an extra header file going to slow down compilation?

**Comment by fredrik-johansson** [comment]
This looks OK in general except that flint_types.h doesn't look necessary (just put the definitions in flint.h).

We should document somewhere that the user needs to include <stdio.h> before any FLINT headers to use functions with file IO.

Also, this PR seems to sneak back FLINT_NOSTDIO and FLINT_NOSTDARG unannounced ;-) If we want to require user inclusion of <stdio.h> for FILE functions and <stdarg.h> for vprintf functions, then I think it'd be better to just do so instead of having conditional logic for it?

**Comment by albinahlback** [comment]
> This is not a good idea. When you have computed a huge object that pushes the limits of your RAM, you want to be able to write it to a file without building a string (which may be even larger than the original object) in memory.

Yeah, that's true. Then I take it back haha

File: src/flint.h (hunk 1 above)
Comment (albinahlback) [reply src/flint.h:29]: Yeah, probably. However, not by much.

**Comment by albinahlback** [comment]
> This looks OK in general except that flint_types.h doesn't look necessary (just put the definitions in flint.h).

Alright, I'll fix that!

> We should document somewhere that the user needs to include <stdio.h> before any FLINT headers to use functions with file IO.
> 
> Also, this PR seems to sneak back FLINT_NOSTDIO and FLINT_NOSTDARG unannounced ;-) If we want to require user inclusion of <stdio.h> for FILE functions and <stdarg.h> for vprintf functions, then I think it'd be better to just do so instead of having conditional logic for it?


Yeah, but for now `stdio.h` and `stdarg.h` should still be opt-out. Unless you accidentally define the macros, it should not matter.

But it's only a draft as well. I mainly added it for myself so I get warnings and error. I mainly want to structure everything so that binaries won't differ if the user compile with or without `FLINT_NOSTDIO`.

**Comment by albinahlback** [comment]
> > This is not a good idea. When you have computed a huge object that pushes the limits of your RAM, you want to be able to write it to a file without building a string (which may be even larger than the original object) in memory.
> 
> Yeah, that's true. Then I take it back haha

However, one could set a buffer on its own at, say, 5kB. So as soon as it knows it will probably surpass that limit, it will push to the file stream. Sounds better?

**Comment by fredrik-johansson** [comment]
> However, one could set a buffer on its own at, say, 5kB. So as soon as it knows it will probably surpass that limit, it will push to the file stream. Sounds better?

But again, this is presumably what ``fprintf`` already does. So what do we gain by doing it ourselves?

**Comment by albinahlback** [comment]
> > However, one could set a buffer on its own at, say, 5kB. So as soon as it knows it will probably surpass that limit, it will push to the file stream. Sounds better?
> 
> But again, this is presumably what `fprintf` already does. So what do we gain by doing it ourselves?

Hey, don't criticize my ideas! ;-)

**Comment by albinahlback** [comment]
Now the binaries should be the same, regardless of opt-out or not.

**Comment by albinahlback** [comment]
Note that currently one has to define their own CPPFLAGS in order to opt-out. I plan to introduce an option for this later on, but this is enough for this PR.

**Comment by albinahlback** [comment]
@fredrik-johansson I'll wait for your opinion before I merge.

I'll iterate this once again: Removing `stdio.h` and `stdarg.h` is totally opt-out. The default (for now at least) should that they are included.

**Comment by albinahlback** [comment]
Going to merge when all non-Windows checks pass.

**Pull request merged**